runtime/src/main/scala/com/armanbilge/vilcacora/runtime/Interpreter.scala

```diff
@@ -0,0 +1,305 @@
+      for {
+        _ <- copyInputsToMemory(inputs, model, memoryMap)
+
+        // The core execution loop. It executes each operation and then yields control
+        // to the Cats Effect runtime to maintain application responsiveness.
+        // This is the desired structure.
+        _ <- model.operations.traverse_ { op =>
+          executeOperation(op, memoryMap, model) <* IO.cede
+        }
+
+        results <- copyOutputsFromMemory(model, memoryMap)
+      } yield results
+    }
+
+  /** A `Resource` that manages the lifecycle of all tensor memory for the graph. */
+  private def memoryResource(model: ModelIR): Resource[IO, MemoryMap] = {
+    val allocAll = IO {
+      model.allocations.map { case (name, allocation) =>
+        val totalBytes = (allocation.shape.product * allocation.dataType.sizeInBytes).toUSize
+        val ptr = stdlib.malloc(totalBytes)
+        if (ptr == null) throw new OutOfMemoryError(s"Failed to allocate tensor '$name'")
+        allocation.initialData.foreach(data =>
+          memcpy(ptr, data.at(0).asInstanceOf[Ptr[Byte]], data.length.toUSize),
+        )
+        name -> ptr
+      }
+    }
+    Resource.make(allocAll)(memoryMap => IO(memoryMap.values.foreach(stdlib.free))).map(_.toMap)
+  }
+
+  /** Executes a single operation by dispatching to the appropriate handler. */
+  private def executeOperation(op: Operation, memory: MemoryMap, model: ModelIR): IO[Unit] =
+    op match {
+      case op: Operation.SVMClassifier => handleSvmClassifier(op, memory, model)
+      case op: Operation.Add => handleElementWise(op, memory, model)(_ + _)
+      case op: Operation.Mul => handleElementWise(op, memory, model)(_ * _)
+      case op: Operation.Cast => handleCast(op, memory, model)
+      case other =>
+        IO.raiseError(
+          new NotImplementedError(s"Operation not implemented: ${other.getClass.getSimpleName}"),
+        )
+    }
+
+  // --- Operation Handlers ---
+
+  private def handleElementWise(op: Operation, memory: MemoryMap, model: ModelIR)(
+      f: (Float, Float) => Float,
+  ): IO[Unit] = IO {
+    val outputAllocation = model.allocations(op.outputs.head)
+    val count = outputAllocation.shape.product
+
+    val inputA = memory(op.inputs(0)).asInstanceOf[Ptr[CFloat]]
+    val inputB = memory(op.inputs(1)).asInstanceOf[Ptr[CFloat]]
+    val output = memory(op.outputs(0)).asInstanceOf[Ptr[CFloat]]
+
+    for (i <- 0 until count)
+      !(output + i) = f(!(inputA + i), !(inputB + i))
+  }
+
+  private def handleCast(op: Operation.Cast, memory: MemoryMap, model: ModelIR): IO[Unit] = IO {
+    val inputAlloc = model.allocations(op.input)
+    val outputAlloc = model.allocations(op.output)
+
+    val inputPtr = memory(op.input)
+    val outputPtr = memory(op.output)
+    val count = inputAlloc.shape.product
+
+    (inputAlloc.dataType, outputAlloc.dataType) match {
+      case (from, to) if from == to =>
+        val _ = memcpy(outputPtr, inputPtr, (count * from.sizeInBytes).toUSize)
```

Comment:
This case can be optimized away: a cast between the same datatypes is a no-op, so we should avoid copying memory for performance.